src/cart/crt_swim.c

```diff
@@ -94,27 +103,73 @@ static void crt_swim_srv_cb(crt_rpc_t *rpc_req)
 {
 	struct crt_grp_priv	*grp_priv = crt_gdata.cg_grp->gg_srv_pri_grp;
 	struct crt_swim_membs	*csm = &grp_priv->gp_membs_swim;
+	struct swim_context	*ctx = csm->csm_ctx;
 	struct crt_rpc_swim_in	*rpc_swim_input = crt_req_get(rpc_req);
-	int			 rc;
+	struct crt_rpc_swim2_out *rpc_swim_output = crt_reply_get(rpc_req);
+	swim_id_t		 self_id = swim_self_get(ctx);
+	int			 rc = -ENOTCONN;
 
 	D_ASSERT(crt_is_service());
 
-	D_DEBUG(DB_TRACE, "incoming message %lu <= %lu\n",
-		swim_self_get(csm->csm_ctx), rpc_swim_input->src);
+	D_TRACE_DEBUG(DB_TRACE, rpc_req,
+		"incoming opc %#x with %zu updates %lu <= %lu\n",
+		rpc_req->cr_opc, rpc_swim_input->upds.ca_count,
+		self_id, rpc_swim_input->src);
+
+	if (self_id != SWIM_ID_INVALID) {
+		rc = swim_parse_message(ctx, rpc_swim_input->src,
+					rpc_swim_input->upds.ca_arrays,
+					rpc_swim_input->upds.ca_count);
+		if (rc == -ESHUTDOWN)
+			swim_self_set(ctx, SWIM_ID_INVALID);
+		else if (rc)
+			D_ERROR("swim_parse_message() failed rc=%d\n", rc);
+	}
+
+	if (rpc_req->cr_opc & 0xFFFF) {
```

Comment:
What is this 0xFFFF?

Reply:
This mean opc index not zero. This is how I check the ping is with reply or not.